### ISSUE TITLE
`Effloglog` class - justification for prior mean `mu0` needed.

### ISSUE THREAD

**Issue by wwojciech** [Roche/crmPack#195]
The `Effloglog` class is used for the linear log-log efficacy model using pseudo  data prior. This efficacy log-log model is given as
```
eff_i = theta1 + theta2 * log(log(dose_i)) + epsilon_i,
```
where `eff_i` is the efficacy response for subject `i` and `dose_i` is the dose level treated for subject `i`.

This issue is to clarify the following line of the code (and its consequences or assumptions) in `Effloglog` construction function, that is used to compute the value of posterior mean of the parameters `theta1` (intercept) and `theta2` (slope) of the efficacy model. It is:
```
mu <- MASS::ginv(Q0 + t(X) %*% X) %*% (Q0 %*% mu0 + t(X) %*% w2),
```
where
`Q0 := t(X0) %*% X0)` is the precision matrix based on pseudo data only (prior precision matrix), `X0` - design matrix with pseudo dose levels only,
`X` - design matrix with observed dose levels only (that correspond to no DLT only),
`mu0 = c(theta1, theta2)`, where `theta1`, `theta2` are the OLS estimates in the efficacy linear model, based on combined (pseudo + observed and no DLT) data,
`w2` - observed efficacy responses only (no DLT).

The standard formula for the posterior mean `m` in Bayesian Linear Regression (see e.g. https://en.wikipedia.org/wiki/Bayesian_linear_regression) is
```
m = (Q0 + X^T * X)^{-1}  (Q0 * m0 + X^T * X * Bhat),
```
where
`Bhat` is the standard OLS estimator in the linear model, which is based on data included in the design matrix X.
`m0`- prior mean,
`Q0` - prior precision matrix.

Now, given that `mu0` is the OLS estimator of `theta1`, `theta2` that is based on combined (pseudo + observed, no DLT) data (as it is in current implementation), then `mu` (as coded above) is equivalent to theoretical `m`, which is fine. However, I am concerned that the prior mean `mu0` is not based only on observed data, but rather on full data (including observed), which I think should not be the case. I believe that `mu0` should be based on prior data only. If it so, the current code for `mu` is wrong, and it should be:
```
mu0 <- coefficients(lm(eff ~ log(log(eff_dose)))) # eff, eff_dose - pseudo data only
Bhat <- coefficients(lm(w2 ~ log(log(x2)))) # w2 (efficacy), x2 (dose) - observed data only (no DLT)
mu <- MASS::ginv(Q0 + t(X) %*% X) %*% (Q0 %*% mu0 + t(X) %*% X %*% Bhat
```
or simply
```
X_full <- rbind(X0, X)
w <- c(eff, w2)
mu <- MASS::ginv(t(X_full) %*% X_full) %*% X_full %*% eff_dose_full
```
as these two formulas for `mu` are equivalent. And then clearly, 
```
mu = OLS estimates for `theta1`, `theta2` that are based on combined (pseudo + observed, no DLT),
```
which makes a lot of sense.

**Comment by Winnie-Yeung** [comment]
yes as long as mu0 is estimated only from psuedo data and mu is from pseudo and observed data, then it is ok. Please adapt accordingly and please check if this runs too in the new adaptation.

**Comment by wwojciech** [comment]
Thank you @Winnie-Yeung

Ok, so in this case, I have 2 more questions:

1. If `mu0` is an OLS estimator based on pseudo data only - I think that we should allow for more observations than just 2, as we discussed. Otherwise, it is just a perfect fit, and not any benefits from the statistical modeling.
2. The estimates of the main model coefficients: `theta1` (intercept) and `theta2` (slope) are stored in the model class slots. With the existing implementation these coefs are estimated with OLS based on combined data (pseudo + observed). Is this OK , or these should be estimated on observed only? I think that it is OK, but please just confirm.

@danielinteractive - FYI

Thanks!

**Comment by Winnie-Yeung** [comment]
Hi
 Does this have an error only when using 2 doses?. The idea here  is that there is a pre-step MCMC here before we obtain the prior mu0. The user will only provide the pseudo data for 2 dose (i.e. 2 betas) . Then the MCMC will allow to estimate all betas in doses other than these 2 doses . Then all together we will get all the prior value of betas for all doses i.e.  the mu0 representing with all doses.  it is like using 2 pseudo data from 2 dose to obtain some 'observed/pseudo data' for other doses. Then all together using all pseudo data in all dose to obtained mu0 . Then with actually observed data + all these pseudo data to obtained mu the posterior estimate mu
This was discussed in 
![image](https://user-images.githubusercontent.com/104487616/165495114-254541fa-ff21-4685-bf8f-1a9deaa21855.png)

If user provide more than 2 pseudo  doses here, it will certainly helps the MCMC for prior mu0 but this may not works in other parts of the program?

I am not sure if this answer the question.

Many thanks.
BR
Winnie

**Comment by wwojciech** [comment]
You can estimate coefs is simple linear regression model by OLS based on 2 observations only, it will be just a perfect fit, but no any errors. But, then why do we need regression model and the OLS for the prior, if you assume you have only 2 observations for prior? If you have 2 observations only, then a simple analytical formula for `theta1`, `theta2` - resulting from the equation of a straight line that crosses to two fixed (2-dim) points - no any modeling here. Hence my question - if you decided to use regression model for prior, then I would assume that you expect that there might be more than 2 observations. Otherwise it does not make much sense.  If I provide more than 2 pseudo observations - it should work fine, the only object that will change is `vecY` if no observed data is available, as the length of this vector is equal to the number of pseudo eff/doses available. Is this ok?

What about other points?



What about the other q

**Comment by Winnie-Yeung** [comment]
I think I forwarded you the wrong paper as I get confused this is an earlier work than the publication. Yes the purpose of  the 2 observations/2 pseudo data is actually estimate by only the simple linear model to get theta1 theta2 from a line. So either as you suggest, we may have to notify and allow the user to use more than 2 dose and have to fix the code when using more than 2 doses. So I think this option should also work even only 2 observations is provided? (or could we specific too if only 2 does is provide then fit simple linear model?) 
the second question is please use combined data (pseudo + observed).

**Comment by wwojciech** [comment]
Hi

I wanted to also ask about the `mcmc` method that is used to obtain the posterior samples for this `Effloglog` model. Here it the metod you wrote (after some bug corrections and minor cosmetic changes from my side):

```
setMethod(
  f = "mcmc",
  signature = signature(
    data = "DataDual",
    model = "Effloglog",
    options = "McmcOptions"
  ),
  definition = function(data, model, options, ...) {

    # Eecide whether we sample from the prior or not.
    from_prior <- data@nObs == 0L

    model <- update(object = model, data = data)
    sample_size <- sampleSize(options)

    if (model@use_fixed) {
      nu <- model@nu
      nu_samples <- rep(nu, sample_size)
    } else {
      nu_samples <- rgamma(sample_size, shape = model@nu["a"], rate = model@nu["b"])
      nu <- mean(nu_samples)
    }

    # Sample from the (asymptotic) bivariate normal prior for theta1 and theta2.
    tmp <- mvtnorm::rmvnorm(
      n = sample_size,
      mean = c(model@theta1, model@theta2),
      sigma = solve(nu * (model@Q))
    )

    samples <- list(
      theta1 = tmp[, 1],
      theta2 = tmp[, 2],
      nu = nu_samples
    )

    Samples(
      data = samples,
      options = options
    )
  }
)
```

1. Why is `from_prior` not used within this function?
2. Assuming that it should be used, shouldn't be 
```
from_prior <- sum(!data@y) == 0L. 
```
as we want to limit to no DLT only and not to just any observed responses?

Thank you in advance!

**Comment by Winnie-Yeung** [comment]
Yes it should be used and please update as you suggested . Thank you

**Comment by wwojciech** [comment]
Thanks!

So how should it be used then? (not asking for the initialisation of `from_prior` as this is already clear).

W.

**Comment by Winnie-Yeung** [comment]
I have the same question as you do. I am not sure why this line is there. I had a quick look too at the initial version I did 8 years ago and I do not see this so I am not 100% sure how this will be used.  This may cause errors within this line when this function is used together with other functions (e.g. other new function developed in the last 7 years)  within the package?? ?

**Comment by wwojciech** [comment]
Thanks. I see. No, this line does not cause any issues with the other functions.

**Comment by danielinteractive** [comment]
@Winnie-Yeung @wwojciech I would guess that when we want to sample from the prior (because data is empty) then can just skip the `update(object = model, data = data)` line. Although update probably does the right thing in that case anyway, so then we don't need to define `from_prior` at all.
Important would still be that we can sample from the model prior in case there is no data.

**Comment by wwojciech** [comment]
Thanks @danielinteractive So for now - I will just remove `from_prior` variable (as I did), without any further modifications.
